src/cargo/ops/cargo_rustc/links.rs

```diff
@@ -15,11 +15,23 @@ pub fn validate(deps: &PackageSet) -> CargoResult<()> {
         };
         match map.get(&lib) {
             Some(previous) => {
-                return Err(human(format!("native library `{}` is being linked \
-                                          to by more than one package, and \
-                                          can only be linked to by one \
-                                          package\n\n  {}\n  {}",
-                                         lib, previous, dep.package_id())))
+                if previous.name() == dep.package_id().name() {
+                  return Err(human(format!("native library `{}` is being \
+                                            linked to by more than one \
+                                            version of the same package, and \
+                                            can only be linked \
+                                            to by one package. Try updating or \
+                                            pinning your dependencies to make \
+                                            the versions of this package match \
```

Comment:
Perhaps something along the lines of:


> native library `foo` is being linked to by more than one version of the same package, but it can only be linked once; try updating dependencies to ensure that this package only shows up once